### ISSUE TITLE
azuread_group: Support owners and members management 

### ISSUE THREAD

**Issue by perbergland** [terraform-providers/terraform-provider-azuread#36]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* azuread_group

### Potential Terraform Configuration

```hcl
resource "azuread_group" "mygroup" {
name= "mygroup"
members = ["${azuread_member.member1.id}","guid-guid-guid-guid"]
owners = ["${azuread_member.member1.id}","guid-guid-guid-guid"]
}

# or maybe this is better to allow a mixture of managed and unmanaged owners/members

resource "azuread_group_owner" "managed_owner" {
group = "${azuread_group.mygroup.id}"
owner = "${azuread_member.member1.id}"
}

resource "azuread_group_member" "managed_member" {
group = "${azuread_group.mygroup.id}"
member = "${azuread_member.member1.id}"
}
```

**Comment by ksix** [comment]
The model of mixed managed and unmanaged owners & members is what I need. It also gives the most flexibility.

**Comment by tom-dudley** [comment]
I'm interested in seeing this implemented. I've done a bit of work [here](https://github.com/tom-dudley/terraform-provider-azuread/tree/ds_r_azuread_group_membership) which implements group membership (not ownership yet) using UPNs:

```
resource "azuread_group" "mygroup" {
    name = "MyGroup"
    members = ["foo.bar@example.com"]
 }
```

It's still a WIP and lacking tests, but does currently work for membership.

Given my use case involves creating new groups and adding pre-existing users, the UPNs seem more maintainable than guids, although I can see the following may be a better option:

```
data "azuread_user" "foo_bar" {
    user_principal_name = "foo.bar@example.com"
 }

resource "azuread_group" "mygroup" {
    name = "MyGroup"
    members = ["${azuread_user.foo_bar.id}"]
 }
```

The only downside I see here is having to have a data element for each user, although the loops in `v 0.12` may solve that.

Couple of questions:
* Was the `azuread_member` meant to be `azuread_user`, or are you suggesting a new resource?
* I still don't understand the need for the second proposal with regards to unmanaged users, do you mean for unmanaged (pre-existing) groups? I can see how it'd be helpful in that respect.

**Comment by perbergland** [comment]
Yeah azuread_member was a referral to a user resource which isn’t released yet.

As for the unmanaged users, for my use case I will want to add msi principals to my management groups for azure and sw server through terraform but keep the management of adding or removing other normal users and groups to/from the managed group a manual task

**Comment by ksix** [comment]
@tom-dudley if it is done using a single resource as in your example, then if anyone adds users outside of terraform, the next time it runs it will remove any users not specified in the members list.  Using separate resources allows individual users to be mapped only, without impacting the whole collection.

**Comment by StefanKert** [comment]
Wouldn´t it make sense to have both in place? I get that some will do things manually, but IMO the case that @ksix mentioned is a valid one (at least for us). If I want to make sure that my group does only contain the users that are defined in my terraform file I think it would be good to remove the not defined groupmembers from the group. To me this is similar to infrastructure since IMO we can consider Azure AD Groups as infrastructure. If someone changes something in the server configuration I love that terraform moves this back or at least gives me a hint that the infra is not in the desired state. 

To summarize: I would love to have a possiblity to define desired states for azure_ad_groups and to define which members are part of it.

**Comment by katbyte** [comment]
@StefanKert & @ksix,

Both approaches are valid and I am of the mind we should support both, however i think we may need to discuss this as they do conflict with each other. The question is how best to implement both as well see if we can guard against someone using both.

**Comment by a144944** [comment]
@tom-dudley:
> The only downside I see here is having to have a data element for each user, although the loops in v 0.12 may solve that.

The following snippet should be able to iterate through without needing terraform 12 syntax:

```hcl
# Get user ids
data "azuread_user" "user" {
  count               = "${length(var.users)}"
  user_principal_name = "${element(var.users, count.index)}"
}

resource "azuread_group_member" "group_member" {
  count = "${length(var.users)}"

  principal_id       = "${element(data.azuread_user.user.*.id, count.index)}"
  group_id           = "${var.group_id}"
}
```

**Comment by ksix** [comment]
> @StefanKert & @ksix,
> 
> Both approaches are valid and I am of the mind we should support both, however i think we may need to discuss this as they do conflict with each other. The question is how best to implement both as well see if we can guard against someone using both.

I wouldn't worry too much about people using both. There are other terraform resources which can potentially collide. I'd just make it clear in the docs to use one or the other.
